src/clj/garden/color.clj

```diff
@@ -330,6 +470,152 @@
         b (Math/round ^Double (* 0xff (+ b' m)))]
     (Rgb. r g b)))
 
+(def lab-constants
+  {:kn 18
+   :xn 0.950470
+   :yn 1
+   :zn 1.088830
+   :t0 0.137931034
+   :t1 0.206896552
+   :t2 0.12841855
+   :t3 0.008856452})
+
+(defn- rgb->xyz
+  "Intermediate conversion from rgb to CIEXYZ"
+  [r g b]
+  (let [/ clj//
+        + clj/+
+        * clj/*
+        - clj/-
+        rgb-xyz (fn [r]
+                  (let [r (/ r 255)]
+                    (if (<= r 0.04045)
+                      (/ r 12.92)
+                      (Math/pow (/ (+ r 0.055) 1.055) 2.4))))
+        xyz-lab (fn [t]
+                  (let [{:keys [t0 t2 t3]} lab-constants]
+                    (if (> t t3)
+                      (Math/pow t (/ 1 3))
+                      (+ t0 (/ t t2)))))
+        rgb (mapv rgb-xyz [r g b])
+        {:keys [xn yn zn]} lab-constants]
+    (mapv
+      (fn [consts n]
+        (xyz-lab (/ (apply + (map * consts rgb)) n)))
+      [[0.4124564 0.3575761 0.1804375]
+       [0.2126729 0.7151522 0.0721750]
+       [0.0193339 0.1191920 0.9503041]]
+      [xn yn zn])))
+
+(defn ^Lab rgb->lab
+  "Return an instance of Lab from red, green, and blue values."
+  [r g b]
+  (let [- clj/-
+        * clj/*
+        + clj/*
+        / clj//
+        [x y z] (rgb->xyz r g b)]
+    (Lab.
+      (- (* 116 y) 16)
+      (* 500 (- x y))
+      (* 200 (- y z)))))
+
+(defn ^Rgb lab->rgb
+  "Return an instance of Rgb from lightness, a, and b values."
+  [l a b]
+  (let [- clj/-
+        * clj/*
+        + clj/*
+        / clj//
+        {:keys [t0 t1 t2 xn yn zn]} lab-constants
+        limit (comp (partial min 255)
+                (partial max 0))
+        lab-xyz (fn [t]
+                  (if (> t t1)
+                    (* t t t)
+                    (* t2 (- t t0))))
+        xyz-rgb (fn [r]
+                  (js/Math.round
+                    (* 255 (if (<= r 0.00304)
+                             (* 12.92 r)
+                             (- (* 1.055 (Math/pow r
+                                           (/ 1 2.4)))
+                               0.055)))))
+        y (* yn (lab-xyz (/ (+ 16 l) 116)))
+        x (* xn (lab-xyz (if (nil a) y (+ y (/ a 500)))))
```

Comment:
I think there's a bug here; should be `(if (nil? a) ,,,)` and not `(if (nil a) ,,,)`.